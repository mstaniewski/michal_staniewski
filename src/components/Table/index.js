import React from "react";
import PropTypes from "prop-types";

import Pagination from "components/Pagination";

import withLoader from "hocs/withLoader";

const Table = ({
  fields = [],
  data = [],
  onSort,
  pagination,
  onPageChange
}) => {
  return (
    <>
      <table className="table table-bordered table-hover">
        <thead className="thead-light">
          <tr>
            {fields.map(({ title, column }) => (
              <th
                onClick={() => onSort(column)}
                key={`${title}-heading`}
                scope="col"
              >
                {title}
              </th>
            ))}
          </tr>
        </thead>
        <tbody>
          {!data.length && (
            <tr>
              <td className="pt-4 pb-4 text-center" colspan={fields.length}>
                Sorry. No results found
              </td>
            </tr>
          )}
          {data &&
            data.map(entity => (
              <tr key={`row-${entity.id}`}>
                {fields.map(({ column, component: Component }) => (
                  <td key={`${column}-${entity.id}`}>
                    {Component ? <Component {...entity} /> : entity[column]}
                  </td>
                ))}
              </tr>
            ))}
        </tbody>
      </table>
      <Pagination onChange={onPageChange} pagination={pagination} />
    </>
  );
};

Table.defaultProps = {
  inputProps: {}
};

Table.propTypes = {
  fields: PropTypes.arrayOf(PropTypes.shape({})).isRequired,
  data: PropTypes.arrayOf(PropTypes.shape({})).isRequired,
  isFetching: PropTypes.bool.isRequired
};

export { default as TableActions } from "./components/TableActions";

export default withLoader(({ isFetching }) => isFetching)(Table);
