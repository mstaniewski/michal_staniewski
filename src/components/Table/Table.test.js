import React from "react";
import { shallow, mount } from "enzyme";
import Table, { TableActions } from "./index";

const TEST_TABLE_PROPS = {
  fields: [
    { title: "Id", column: "id" },
    { title: "Name", column: "name" },
    { title: "Species", column: "species" },
    { title: "Gender", column: "gender" },
    { title: "Homeworld", column: "homeworld" },
    {
      title: "Actions",
      component: ({ id }) => <TableActions onDelete={() => {}} />
    }
  ],
  data: [],
  isFetching: false
};

describe("<Table />", () => {
  it("should match snapshot", () => {
    const wrapper = shallow(<Table {...TEST_TABLE_PROPS} />);
    expect(wrapper).toMatchSnapshot();
  });

  it("should render proper number of columns", () => {
    const wrapper = mount(<Table {...TEST_TABLE_PROPS} />);
    expect(wrapper.find("th")).toHaveLength(TEST_TABLE_PROPS.fields.length);
  });

  it("should render proper columns headings", () => {
    const wrapper = mount(<Table {...TEST_TABLE_PROPS} />);
    wrapper.find("th").forEach((node, idx) => {
      expect(node.text()).toEqual(TEST_TABLE_PROPS.fields[idx].title);
    });
  });
});
