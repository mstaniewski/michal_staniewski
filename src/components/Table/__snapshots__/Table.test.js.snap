// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Table /> should match snapshot 1`] = `ShallowWrapper {}`;
